File: scripts/deploy_edge_functions.py
#!/usr/bin/env python3

import os
import json
import base64
import requests
from pathlib import Path

# Supabase 설정
SUPABASE_PROJECT_ID = "lgebgddeerpxdjvtqvoi"
SUPABASE_URL = f"https://{SUPABASE_PROJECT_ID}.supabase.co"
SUPABASE_ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImxnZWJnZGRlZXJweGRqdnRxdm9pIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NDkxOTc2MDksImV4cCI6MjA2NDc3MzYwOX0.NZxHOwzgRc-Vjw60XktU7L_hKiIMAW_5b_DHis6qKBE"
SUPABASE_SERVICE_ROLE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImxnZWJnZGRlZXJweGRqdnRxdm9pIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc0OTE5NzYwOSwiZXhwIjoyMDY0NzczNjA5fQ.X-8mGiWHKUKaZW4ZrtDXNUhISTAtZFZGsreB5peGgbQ"
FINNHUB_API_KEY = "d11du61r01qu0d0fu8v0d11du61r01qu0d0fu8vg"

# Management API URL
MANAGEMENT_API_URL = "https://api.supabase.com"

# 프로젝트 루트 디렉토리
PROJECT_ROOT = Path(__file__).parent.parent

def get_function_code(function_name):
    """Edge Function 코드 읽기"""
    function_path = PROJECT_ROOT / "supabase" / "functions" / function_name / "index.ts"
    if function_path.exists():
        return function_path.read_text()
    else:
        print(f"❌ Function file not found: {function_path}")
        return None

def create_edge_function_via_db():
    """데이터베이스를 통해 Edge Function 메타데이터 생성"""
    import psycopg2
    
    # Supabase 데이터베이스 연결
    conn_str = f"postgresql://postgres.{SUPABASE_PROJECT_ID}:{SUPABASE_SERVICE_ROLE_KEY}@aws-0-us-west-1.pooler.supabase.com:5432/postgres"
    
    try:
        conn = psycopg2.connect(conn_str)
        cur = conn.cursor()
        
        # Edge Functions 테이블이 있는지 확인하고 생성
        cur.execute("""
            CREATE TABLE IF NOT EXISTS edge_functions (
                id UUID DEFAULT gen_random_uuid() PRIMARY KEY,
                name TEXT UNIQUE NOT NULL,
                slug TEXT UNIQUE NOT NULL,
                created_at TIMESTAMPTZ DEFAULT NOW(),
                updated_at TIMESTAMPTZ DEFAULT NOW()
            );
        """)
        
        # Functions 등록
        functions = ["market-data", "trading-signals", "portfolio-management"]
        
        for func_name in functions:
            cur.execute("""
                INSERT INTO edge_functions (name, slug)
                VALUES (%s, %s)
                ON CONFLICT (name) DO UPDATE
                SET updated_at = NOW();
            """, (func_name, func_name))
            print(f"✅ Registered function: {func_name}")
        
        conn.commit()
        cur.close()
        conn.close()
        
        print("\n✅ All functions registered in database!")
        
    except Exception as e:
        print(f"❌ Database error: {e}")

def test_edge_function(function_name, test_data):
    """Edge Function 테스트"""
    url = f"{SUPABASE_URL}/functions/v1/{function_name}"
    
    headers = {
        "Authorization": f"Bearer {SUPABASE_ANON_KEY}",
        "Content-Type": "application/json"
    }
    
    print(f"\n🧪 Testing {function_name}...")
    print(f"URL: {url}")
    print(f"Data: {json.dumps(test_data, indent=2)}")
    
    try:
        response = requests.post(url, headers=headers, json=test_data)
        print(f"Status: {response.status_code}")
        
        if response.status_code == 200:
            print(f"✅ Success: {response.json()}")
        else:
            print(f"❌ Error: {response.text}")
            
    except Exception as e:
        print(f"❌ Request failed: {e}")

def create_deployment_package():
    """배포 패키지 생성"""
    print("📦 Creating deployment package...\n")
    
    deployment_dir = PROJECT_ROOT / "supabase-deployment"
    deployment_dir.mkdir(exist_ok=True)
    
    # .env.local 파일 생성
    env_content = f"""# Supabase Edge Functions Environment Variables
SUPABASE_URL={SUPABASE_URL}
SUPABASE_ANON_KEY={SUPABASE_ANON_KEY}
SUPABASE_SERVICE_ROLE_KEY={SUPABASE_SERVICE_ROLE_KEY}
FINNHUB_API_KEY={FINNHUB_API_KEY}
"""
    
    (deployment_dir / ".env.local").write_text(env_content)
    print("✅ Created .env.local")
    
    # import_map.json 생성
    import_map = {
        "imports": {
            "std/": "https://deno.land/std@0.168.0/",
            "supabase": "https://esm.sh/@supabase/supabase-js@2"
        }
    }
    
    (deployment_dir / "import_map.json").write_text(json.dumps(import_map, indent=2))
    print("✅ Created import_map.json")
    
    # 각 함수별 디렉토리 생성
    functions = ["market-data", "trading-signals", "portfolio-management"]
    
    for func_name in functions:
        func_dir = deployment_dir / func_name
        func_dir.mkdir(exist_ok=True)
        
        # 함수 코드 복사
        source_code = get_function_code(func_name)
        if source_code:
            (func_dir / "index.ts").write_text(source_code)
            print(f"✅ Copied {func_name}/index.ts")
    
    # 배포 스크립트 생성
    deploy_script = """#!/bin/bash
# Supabase Edge Functions Local Deployment Script

echo "🚀 Deploying Edge Functions locally..."

# Supabase CLI로 로컬 개발 서버 시작
supabase start

# Functions 서빙
supabase functions serve --env-file .env.local

echo "✅ Edge Functions are now running locally!"
echo "Test URLs:"
echo "  - http://localhost:54321/functions/v1/market-data"
echo "  - http://localhost:54321/functions/v1/trading-signals"
echo "  - http://localhost:54321/functions/v1/portfolio-management"
"""
    
    (deployment_dir / "deploy-local.sh").write_text(deploy_script)
    os.chmod(deployment_dir / "deploy-local.sh", 0o755)
    print("✅ Created deploy-local.sh")
    
    print(f"\n📁 Deployment package created at: {deployment_dir}")
    print("\nTo deploy locally:")
    print(f"  cd {deployment_dir}")
    print("  ./deploy-local.sh")

def main():
    print("🚀 Supabase Edge Functions Deployment\n")
    
    # 1. 배포 패키지 생성
    create_deployment_package()
    
    # 2. 데이터베이스에 함수 등록 시도
    print("\n📝 Registering functions in database...")
    create_edge_function_via_db()
    
    # 3. 테스트 데이터로 함수 테스트
    print("\n🧪 Testing Edge Functions...")
    
    # market-data 테스트
    test_edge_function("market-data", {
        "action": "quote",
        "symbol": "AAPL"
    })
    
    # trading-signals 테스트
    test_edge_function("trading-signals", {
        "symbol": "AAPL",
        "strategy": "jesse_livermore",
        "timeframe": "D"
    })
    
    # portfolio-management 테스트
    test_edge_function("portfolio-management", {
        "action": "calculate_performance",
        "portfolioId": "test-portfolio-id"
    })
    
    print("\n✅ Deployment process completed!")
    print("\n📌 Important Notes:")
    print("1. Edge Functions need to be deployed via Supabase CLI or Dashboard")
    print("2. Use 'supabase functions deploy <function-name>' to deploy")
    print("3. Set secrets with 'supabase secrets set KEY=value'")
    print("\n🔗 Supabase Dashboard:")
    print(f"   https://app.supabase.com/project/{SUPABASE_PROJECT_ID}/functions")

if __name__ == "__main__":
    main()

File: scripts/deploy-edge-functions.js
#!/usr/bin/env node

const https = require('https');
const fs = require('fs');
const path = require('path');

// Supabase 프로젝트 정보
const SUPABASE_PROJECT_ID = 'lgebgddeerpxdjvtqvoi';
const SUPABASE_SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImxnZWJnZGRlZXJweGRqdnRxdm9pIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc0OTE5NzYwOSwiZXhwIjoyMDY0NzczNjA5fQ.X-8mGiWHKUKaZW4ZrtDXNUhISTAtZFZGsreB5peGgbQ';

// Edge Functions 코드 읽기
const functionsDir = path.join(__dirname, '..', 'supabase', 'functions');

const functions = [
  {
    name: 'market-data',
    path: path.join(functionsDir, 'market-data', 'index.ts')
  },
  {
    name: 'trading-signals',
    path: path.join(functionsDir, 'trading-signals', 'index.ts')
  },
  {
    name: 'portfolio-management',
    path: path.join(functionsDir, 'portfolio-management', 'index.ts')
  }
];

// Supabase Management API를 통한 Edge Function 배포
async function deployFunction(functionName, functionCode) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify({
      name: functionName,
      slug: functionName,
      body: functionCode,
      verify_jwt: true,
      import_map: true
    });

    const options = {
      hostname: `${SUPABASE_PROJECT_ID}.supabase.co`,
      port: 443,
      path: '/functions/v1/deploy',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
        'Content-Length': Buffer.byteLength(data)
      }
    };

    const req = https.request(options, (res) => {
      let responseData = '';

      res.on('data', (chunk) => {
        responseData += chunk;
      });

      res.on('end', () => {
        if (res.statusCode === 200 || res.statusCode === 201) {
          console.log(`✅ Successfully deployed ${functionName}`);
          resolve(JSON.parse(responseData));
        } else {
          console.error(`❌ Failed to deploy ${functionName}: ${res.statusCode}`);
          console.error(responseData);
          reject(new Error(`Deployment failed: ${responseData}`));
        }
      });
    });

    req.on('error', (error) => {
      console.error(`❌ Error deploying ${functionName}:`, error);
      reject(error);
    });

    req.write(data);
    req.end();
  });
}

// Secrets 설정
async function setSecret(name, value) {
  return new Promise((resolve, reject) => {
    const data = JSON.stringify({
      name: name,
      value: value
    });

    const options = {
      hostname: `${SUPABASE_PROJECT_ID}.supabase.co`,
      port: 443,
      path: '/functions/v1/secrets',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${SUPABASE_SERVICE_ROLE_KEY}`,
        'Content-Length': Buffer.byteLength(data)
      }
    };

    const req = https.request(options, (res) => {
      let responseData = '';

      res.on('data', (chunk) => {
        responseData += chunk;
      });

      res.on('end', () => {
        if (res.statusCode === 200 || res.statusCode === 201) {
          console.log(`✅ Successfully set secret ${name}`);
          resolve();
        } else {
          console.error(`❌ Failed to set secret ${name}: ${res.statusCode}`);
          reject(new Error(`Secret setting failed: ${responseData}`));
        }
      });
    });

    req.on('error', (error) => {
      console.error(`❌ Error setting secret ${name}:`, error);
      reject(error);
    });

    req.write(data);
    req.end();
  });
}

// 메인 배포 함수
async function main() {
  console.log('🚀 Starting Edge Functions deployment...\n');

  try {
    // 1. Secrets 설정
    console.log('📝 Setting up secrets...');
    await setSecret('FINNHUB_API_KEY', 'd11du61r01qu0d0fu8v0d11du61r01qu0d0fu8vg');
    console.log('✅ Secrets configured\n');

    // 2. Edge Functions 배포
    console.log('📦 Deploying Edge Functions...');
    
    for (const func of functions) {
      console.log(`\n📌 Deploying ${func.name}...`);
      
      // 함수 코드 읽기
      const functionCode = fs.readFileSync(func.path, 'utf8');
      
      // 배포
      await deployFunction(func.name, functionCode);
    }

    console.log('\n✅ All Edge Functions deployed successfully!');
    console.log('\n📌 Edge Functions URLs:');
    console.log(`- https://${SUPABASE_PROJECT_ID}.supabase.co/functions/v1/market-data`);
    console.log(`- https://${SUPABASE_PROJECT_ID}.supabase.co/functions/v1/trading-signals`);
    console.log(`- https://${SUPABASE_PROJECT_ID}.supabase.co/functions/v1/portfolio-management`);

  } catch (error) {
    console.error('\n❌ Deployment failed:', error);
    process.exit(1);
  }
}

// 직접 API 호출을 통한 배포 대신 curl 명령어 생성
async function generateDeploymentCommands() {
  console.log('📝 Generating deployment commands...\n');

  // Secrets 설정 명령어
  console.log('# Set secrets:');
  console.log(`curl -X POST https://${SUPABASE_PROJECT_ID}.supabase.co/functions/v1/secrets \\
  -H "Authorization: Bearer ${SUPABASE_SERVICE_ROLE_KEY}" \\
  -H "Content-Type: application/json" \\
  -d '{"name": "FINNHUB_API_KEY", "value": "d11du61r01qu0d0fu8v0d11du61r01qu0d0fu8vg"}'`);

  console.log('\n# Deploy functions:');
  
  for (const func of functions) {
    const functionCode = fs.readFileSync(func.path, 'utf8');
    const escapedCode = functionCode.replace(/'/g, "'\"'\"'");
    
    console.log(`\n# Deploy ${func.name}:`);
    console.log(`curl -X POST https://${SUPABASE_PROJECT_ID}.supabase.co/functions/v1/${func.name} \\
  -H "Authorization: Bearer ${SUPABASE_SERVICE_ROLE_KEY}" \\
  -H "Content-Type: application/json" \\
  -d @- <<'EOF'
{
  "verify_jwt": true,
  "import_map": true,
  "body": ${JSON.stringify(functionCode)}
}
EOF`);
  }
}

// Supabase CLI를 사용한 배포 (권장)
async function deployWithSupabaseCLI() {
  console.log('🚀 Deploying Edge Functions with Supabase CLI...\n');

  const { execSync } = require('child_process');

  try {
    // 프로젝트 링크
    console.log('📌 Linking to Supabase project...');
    execSync(`supabase link --project-ref ${SUPABASE_PROJECT_ID}`, { stdio: 'inherit' });

    // Secrets 설정
    console.log('\n📝 Setting secrets...');
    execSync(`supabase secrets set FINNHUB_API_KEY=d11du61r01qu0d0fu8v0d11du61r01qu0d0fu8vg`, { stdio: 'inherit' });

    // Functions 배포
    console.log('\n📦 Deploying functions...');
    for (const func of functions) {
      console.log(`\nDeploying ${func.name}...`);
      execSync(`supabase functions deploy ${func.name}`, { stdio: 'inherit' });
    }

    console.log('\n✅ All functions deployed successfully!');

  } catch (error) {
    console.error('❌ Deployment failed:', error.message);
    process.exit(1);
  }
}

// 실행 옵션 선택
if (process.argv.includes('--cli')) {
  deployWithSupabaseCLI();
} else if (process.argv.includes('--commands')) {
  generateDeploymentCommands();
} else {
  console.log('Usage:');
  console.log('  node deploy-edge-functions.js --cli      # Deploy using Supabase CLI (recommended)');
  console.log('  node deploy-edge-functions.js --commands # Generate curl commands');
  console.log('\nNote: Make sure you have Supabase CLI installed for --cli option');
}